File: problem/bry/C++/6. Zigzag Conversion.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        string ans = "";
        int n = numRows-1;
        bool tmp = false;
        if(n)
            for(int i = 0 ; i < numRows ; ++i){
                tmp = false;
                for(int j = i; j < s.length();){
                    ans+= s[j];
                    j += tmp ? i*2 : ((n-i)*2)?(n-i)*2:n*2;
                    tmp = (i%n)? !tmp : tmp;
                }
            }
        else
            ans = s;
        return ans;
    }
};

File: problem/bry/Kotlin/6. Zigzag Conversion.kt
// 97.85 98.77
class Solution {
    fun convert(s: String, numRows: Int): String {
        val ans = StringBuilder()
        val n = s.length
        val tmp = if(numRows == 1) 1 else (numRows-1).shl(1)
        for(i in 0 until numRows){
            var k = i + (numRows-i-1).shl(1)
            for(j in i until n step tmp){
                ans.append(s[j])    
                if(k != i && k != tmp && k < n){
                    ans.append(s[k])
                    k += tmp
                }
            }
        }
        return ans.toString()
    }
}